tests/conftest.py

```diff
@@ -119,11 +118,119 @@ def generate_tests(self, smoke_test_data, random_test_data=[]):
         _: list
             Tests.
         """
-        smoke_tests = [
-            pytest.param(*data.values(), marks=smoke) for data in smoke_test_data
-        ]
-        random_tests = [pytest.param(*data, marks=random) for data in random_test_data]
-        return smoke_tests + random_tests
+        tests = []
+        if smoke_test_data:
+            smoke_tests = [
+                pytest.param(*data.values(), marks=pytest.mark.smoke)
+                for data in smoke_test_data
+            ]
+            tests += smoke_tests
+        if random_test_data:
+            random_tests = [
+                pytest.param(*data.values(), marks=pytest.mark.random)
+                if isinstance(data, dict)
+                else pytest.param(*data, marks=pytest.mark.random)
+                for data in random_test_data
+            ]
+            tests += random_tests
+        return tests
+
+    def _log_exp_composition_data(
+        self, space, num_samples=100, max_n=10, num_n=5, **kwargs
+    ):
+        """Generate Data that checks for log and exp are inverses. Specifically
+
+            :math: `Exp_{base_point}(Log_{base_point}(point)) = point`
+
+
+        Parameters
+        ----------
+        space : cls
+            Manifold class that upon which metric is present.
```

Comment:
NIT: missing verb?

Reply:
does new doc string work? not sure If I completely understand what I mean